### PR TITLE
[FSDP][Feature] TIS implementation

### PR DESCRIPTION
**Pull request by MrWhitezz**
This pull request adds support for Token-level Importance Sampling (TIS) to the FSDP actor training loop, which implements #298. The main changes include updating data packing/unpacking to handle the new `rollout_log_probs` field, integrating TIS logic into the policy loss calculation, and reporting relevant TIS metrics. These updates enable off-policy correction using importance sampling during training.

**TIS integration and policy loss calculation:**

* Added TIS logic to the `train` method in `actor.py`, including calculation of importance weights (`tis`), clipping, and adjustment of the policy loss when `use_tis` is enabled.
* Reported TIS metrics (`tis`, `tis_clipfrac`) in the training output when TIS is used.

**Data packing and unpacking updates:**

* Updated the `pack_sequences` function in `data_packing.py` to accept and pack `rollout_log_probs` alongside other rollout data. 
* Modified the `unpack_sequences` function to correctly unpack and slice `rollout_log_probs` for each instance.

Tested with slime/tests/test_fsdp.sh.

### CODE REVIEW

File: slime/backends/fsdp_utils/actor.py [hunk 1]
```diff
@@ -316,13 +317,35 @@ def train(self, rollout_id, rollout_data_ref):
             ppo_kl = old_log_probs.to(device=log_probs.device) - log_probs
             advantages = advantages.to(device=ppo_kl.device)
             pg_loss, pg_clipfrac = compute_policy_loss(ppo_kl, advantages, self.args.eps_clip, self.args.eps_clip_high)
+            
+            # Initialize TIS variables
+            tis = None
+            tis_clipfrac = None
```
Comment (yitianlian) [inline slime/backends/fsdp_utils/actor.py:323]: You can put this after ` if self.args.use_tis`.

Comment (yitianlian) [inline slime/backends/fsdp_utils/actor.py]: You can put this after ` if self.args.use_tis`.

File: slime/backends/fsdp_utils/actor.py [hunk 2]
```diff
@@ -316,13 +317,35 @@ def train(self, rollout_id, rollout_data_ref):
...
+                    isinstance(batch["rollout_log_probs"], torch.Tensor) and 
+                    batch["rollout_log_probs"].numel() > 0
+                    for batch in unpacked_batches
+                ), "rollout_log_probs must be provided as non-empty torch.Tensor for TIS"
+                
+                rollout_log_probs = torch.cat([batch["rollout_log_probs"] for batch in unpacked_batches], dim=0)
+                rollout_log_probs = rollout_log_probs.to(device=log_probs.device)
+                
+                tis = torch.exp(old_log_probs - rollout_log_probs)
+                tis_clip = torch.clamp(tis, min=getattr(self.args, 'tis_clip_low', 0.1), max=getattr(self.args, 'tis_clip', 5.0))
```
Comment (yitianlian) [inline slime/backends/fsdp_utils/actor.py]: `tis_clip` can be set to 2.0 as it's the default value.

File: slime/backends/fsdp_utils/actor.py [hunk 3]
```diff
@@ -316,13 +317,35 @@ def train(self, rollout_id, rollout_data_ref):
...
+                    "rollout_log_probs" in batch and 
+                    isinstance(batch["rollout_log_probs"], torch.Tensor) and 
+                    batch["rollout_log_probs"].numel() > 0
+                    for batch in unpacked_batches
+                ), "rollout_log_probs must be provided as non-empty torch.Tensor for TIS"
+                
+                rollout_log_probs = torch.cat([batch["rollout_log_probs"] for batch in unpacked_batches], dim=0)
+                rollout_log_probs = rollout_log_probs.to(device=log_probs.device)
+                
+                tis = torch.exp(old_log_probs - rollout_log_probs)
```
Comment (yitianlian) [inline slime/backends/fsdp_utils/actor.py:338]: You can also add `ois` metric just like what we do in megaton backend.

File: slime/backends/fsdp_utils/actor.py [hunk 4]
```diff
@@ -349,6 +372,10 @@ def train(self, rollout_id, rollout_data_ref):
 
             if self.args.use_kl_loss:
                 reported["kl_loss"] = kl_loss.detach()
+                
+            if self.args.use_tis and tis is not None:
+                reported["tis"] = sum_of_sample_mean(tis, response_lengths, loss_masks).detach()
```
Comment (yitianlian) [inline slime/backends/fsdp_utils/actor.py:378]: add `ois` metric here.

Comment (Hecate0821) [reply slime/backends/fsdp_utils/actor.py:378]: Resolved

File: slime/backends/fsdp_utils/actor.py (hunk 3 above)
Comment (Hecate0821) [reply slime/backends/fsdp_utils/actor.py:338]: Resolved

File: slime/backends/fsdp_utils/actor.py (hunk 2 above)
Comment (Hecate0821) [reply slime/backends/fsdp_utils/actor.py]: Resolved

File: slime/backends/fsdp_utils/actor.py (hunk 1 above)
Comment (Hecate0821) [reply slime/backends/fsdp_utils/actor.py]: Resolved

**Comment by WWWjiahui** [comment]
Tested on slime/tests/test_qwen3-0.6B_fsdp_colocated_2xGPU.sh. LGTM
截屏2025-10-03 下午5 00 10
